File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.java
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {

        // Step 1: Create dummy node
        ListNode dummy = new ListNode(0);
        dummy.next = head;

        ListNode fast = dummy;
        ListNode slow = dummy;

        // Step 2: Move fast pointer n steps ahead
        for (int i = 0; i < n; i++) {
            fast = fast.next;
        }

        // Step 3: Move both until fast reaches last node
        while (fast.next != null) {
            fast = fast.next;
            slow = slow.next;
        }

        // Step 4: Delete the target node
        slow.next = slow.next.next;

        // Step 5: Return new head
        return dummy.next;
    }
}

File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def removeNthFromEnd(self, head, n):
        """
        :type head: Optional[ListNode]
        :type n: int
        :rtype: Optional[ListNode]
        """
        length = 0
        current = head
        while current != None:
            length +=1
            current = current.next
        k = length - n

        if k == 0:
            return head.next

            
        current = head
        for i in range(k-1):
           
            current = current.next
        current.next = current.next.next

        return head